data/transactions/currency.go

```diff
@@ -0,0 +1,360 @@
+		}
+
+		return balances.Put(record)
+	}
+
+	// Re-configuration and destroying must be done by the manager key.
+	params, err := getParams(balances, cc.ConfigCurrency)
+	if err != nil {
+		return err
+	}
+
+	if header.Sender != params.Manager {
+		return fmt.Errorf("transaction issued by %v, not manager key %v", header.Sender, params.Manager)
+	}
+
+	record, err := balances.Get(cc.ConfigCurrency.Creator)
+	if err != nil {
+		return err
+	}
+
+	record.Currencies = clone(record.Currencies)
+	record.CurrencyParams = cloneParams(record.CurrencyParams)
+
+	if cc.CurrencyParams == (basics.CurrencyParams{}) {
+		// Destroying a currency.  The creator account must hold
+		// the entire outstanding currency amount.
+		if record.Currencies[cc.ConfigCurrency].Amount != params.Total {
+			return fmt.Errorf("cannot destroy currency: holding only %d/%d", record.Currencies[cc.ConfigCurrency].Amount, params.Total)
+		}
+
+		delete(record.Currencies, cc.ConfigCurrency)
+		delete(record.CurrencyParams, cc.ConfigCurrency.Index)
+	} else {
+		// Changing keys in a currency.
+		if !params.Manager.IsZero() {
+			params.Manager = cc.CurrencyParams.Manager
+		}
+		if !params.Reserve.IsZero() {
+			params.Reserve = cc.CurrencyParams.Reserve
+		}
+		if !params.Freeze.IsZero() {
+			params.Freeze = cc.CurrencyParams.Freeze
+		}
+		if !params.Clawback.IsZero() {
+			params.Clawback = cc.CurrencyParams.Clawback
+		}
+
+		record.CurrencyParams[cc.ConfigCurrency.Index] = params
+	}
+
+	return balances.Put(record)
+}
+
+func (ct CurrencyTransferTxnFields) apply(header Header, balances Balances, spec SpecialAddresses, ad *ApplyData) error {
+	// Default to sending from the transaction sender's account.
+	source := header.Sender
+	clawback := false
+
+	if !ct.CurrencySender.IsZero() {
+		// Clawback transaction.  Check that the transaction sender
+		// is the Clawback address for this currency.
+		params, err := getParams(balances, ct.XferCurrency)
+		if err != nil {
+			return err
+		}
+
+		if header.Sender != params.Clawback {
+			return fmt.Errorf("clawback not allowed: sender %v != clawback %v", header.Sender, params.Clawback)
+		}
+
+		// Transaction sent from the correct clawback address,
+		// execute currency transfer from specified source.
+		source = ct.CurrencySender
+		clawback = true
+	}
+
+	if ct.CurrencyAmount > 0 || ct.CurrencyReceiver != (basics.Address{}) {
+		snd, err := balances.Get(source)
+		if err != nil {
+			return err
+		}
+
+		snd.Currencies = clone(snd.Currencies)
+		sndHolding, ok := snd.Currencies[ct.XferCurrency]
+		if !ok {
+			// Allocating a slot for currency (self-transfer of zero amount).
+			// Initialize holding with default Frozen value.
+			if clawback {
+				return fmt.Errorf("cannot allocate currency slot via clawback")
+			}
+
+			params, err := getParams(balances, ct.XferCurrency)
+			if err != nil {
+				return err
+			}
+
+			sndHolding.Frozen = params.DefaultFrozen
+		}
+
+		if sndHolding.Amount < ct.CurrencyAmount {
```

Comment:
nit: might be simpler to use OSub (maybe of a new type)

Reply:
Sure, done.